main.py

```diff
@@ -0,0 +1,183 @@
+
+template_id_mapping = {
+    ("HH", "GB-ENG", "en"): "0c5a4f95-bfa4-4364-9394-8499b4d777d5",
+    ("HH", "GB-WLS", "en"): "0c5a4f95-bfa4-4364-9394-8499b4d777d5",
+    ("HH", "GB-WLS", "cy"): "755d73d1-0cb6-4f2f-95e9-857a2ad071bb",
+    ("HH", "GB-NIR", "en"): "0889cfa1-c0eb-4ba6-93d9-acc41b060152",
+    ("HH", "GB-NIR", "ga"): "0889cfa1-c0eb-4ba6-93d9-acc41b060152",
+    ("HH", "GB-NIR", "eo"): "0889cfa1-c0eb-4ba6-93d9-acc41b060152",
+    ("HI", "GB-ENG", "en"): "71de56dc-f83b-4899-93ab-7fe61e417c2e",
+    ("HI", "GB-WLS", "en"): "71de56dc-f83b-4899-93ab-7fe61e417c2e",
+    ("HI", "GB-WLS", "cy"): "1001ac43-093d-425c-ac7d-68df5147c603",
+    ("HI", "GB-NIR", "en"): "ed1c2e9f-c81e-4cc2-889c-8e0fa1d2ce1b",
+    ("HI", "GB-NIR", "ga"): "ed1c2e9f-c81e-4cc2-889c-8e0fa1d2ce1b",
+    ("HI", "GB-NIR", "eo"): "ed1c2e9f-c81e-4cc2-889c-8e0fa1d2ce1b",
+    ("CE", "GB-ENG", "en"): "4077d2cf-81cd-462d-9065-f227a7c39a8d",
+    ("CE", "GB-WLS", "en"): "4077d2cf-81cd-462d-9065-f227a7c39a8d",
+    ("CE", "GB-WLS", "cy"): "e4a4ebea-fcc8-463b-8686-5b8a7320f089",
+}
+
+data_fields = ("email_address", "display_address", "tx_id", "questionnaire_id")
+
+
+def log_entry(**kwargs):
+    print(json.dumps(kwargs))
+
+
+# pylint: disable=too-many-return-statements
+def notify(request: Request) -> Tuple[str, int]:
+    if not request.method == "POST":
+        # Note that Cloud Functions expect serialized JSON
+        # to correctly log
+        # https://cloud.google.com/functions/docs/monitoring/logging#writing_structured_logs
+        msg = "Method not allowed"
+        log_entry(severity="ERROR", message=msg)
+        return msg, 405
+
+    if not request.json:
+        msg = "Missing notification request data"
+        log_entry(severity="ERROR", message=msg)
+        return msg, 422
+
+    data = request.json["fulfilmentRequest"]
+
+    tx_id = data.get("tx_id")
+    questionnaire_id = data.get("questionnaire_id")
+
+    if not (form_type := data.get("form_type")):
+        msg = "Missing form_type identifier"
+        log_entry(
+            severity="ERROR",
+            message=msg,
+            tx_id=tx_id,
+            questionnaire_id=questionnaire_id,
+        )
+        return msg, 422
+
+    if not (language_code := data.get("language_code")):
+        msg = "Missing language_code identifier"
+        log_entry(
+            severity="ERROR",
+            message=msg,
+            tx_id=tx_id,
+            questionnaire_id=questionnaire_id,
+        )
+        return msg, 422
+
+    if not (region_code := data.get("region_code")):
+        msg = "Missing region_code identifier"
+        log_entry(
+            severity="ERROR",
+            message=msg,
+            tx_id=tx_id,
+            questionnaire_id=questionnaire_id,
+        )
+        return msg, 422
+
+    template_id = template_id_mapping.get(
+        (form_type, region_code, language_code), os.getenv("NOTIFY_TEST_TEMPLATE_ID")
+    )
```

Comment:
Can we reduce this with a loop like:
```py
    required_keys = ("form_type", "region_code", "language_code")
    for required_key in required_keys:
        if not data.get(required_key):
            msg = f"Missing {required_key} identifier"
            log_entry(
                severity="ERROR",
                message=msg,
                tx_id=tx_id,
                questionnaire_id=questionnaire_id,
            )
            return msg, 422

    template_id = template_id_mapping.get(
        (data["form_type"], data["region_code"], data["language_code"]),
        os.getenv("NOTIFY_TEST_TEMPLATE_ID"),
    )
```

Reply:
Done

Reply:
Sorry slightly modification if you agree. Probably useful to report all missing keys.
```py
    required_keys = ("form_type", "region_code", "language_code")
    missing_keys = (key for key in required_keys if not data.get(key))
    if missing_keys:
        msg = f"Missing {', '.join(missing_keys)} identifier(s)"
        raise InvalidRequestError(msg, 422)
```

Reply:
Do you want the moon on a stick as well? 😆

Reply:
Done 👍🏻